Add Asset and Organization resources

jeffgran-dox opened the pull request:
- Adds some new resources that are newer in the API.
- Updates StructuredScope to have the program id so that it can construct its URLs.

rzhade3 commented on lib/hackerone/client/structured_scope.rb:
```diff
@@ -3,24 +3,41 @@
       delegate *DELEGATES, to: :attributes
 
-      def initialize(scope)
+      def initialize(program_id, scope)
```
In keeping with the pattern that was set in the Swag object:

https://github.com/github/hackerone-client/blob/cd44f298bcccbafe8a4e77cbc0ff00d6ee8a9d84/lib/hackerone/client/swag.rb#L9

Let's set this method signature to `scope, program = nil`. This will also allow this change to be backwards compatible. 

```suggestion
      def initialize(scope, program = nil)
```

Here's another callsite that accepts program, for example: https://github.com/github/hackerone-client/blob/cd44f298bcccbafe8a4e77cbc0ff00d6ee8a9d84/lib/hackerone/client/incremental/activities.rb#L11-L12

rzhade3 commented on lib/hackerone/client/structured_scope.rb:
```diff
@@ -3,24 +3,41 @@
+          type: "structured-scope",
+          attributes: attributes
+        }
+        make_put_request("programs/#{@program_id}/structured_scopes/#{id}", request_body: body)
```
```suggestion
        make_put_request("programs/#{@program.id}/structured_scopes/#{id}", request_body: body)
```

rzhade3 commented on lib/hackerone/client/asset.rb:
```diff
@@ -0,0 +1,62 @@
+      end
+
+      def programs
+        relationships.programs[:data].map{ Program.new(_1) }
```
We still technically support Ruby 2.6 for this library, and unfortunately this syntax doesn't work for [versions below 2.7](https://rubyreferences.github.io/rubychanges/2.7.html#numbered-block-parameters).

```suggestion
        relationships.programs[:data].map{ |p| Program.new(p) }
```

rzhade3 commented on lib/hackerone/client/asset.rb:
```diff
@@ -0,0 +1,62 @@
+
+      delegate *DELEGATES, to: :attributes
+
+      def initialize(organization_id, asset)
```
The pattern in other objects of this library is currently `object, additional_parameter`:

```suggestion
      def initialize(asset, organization)
```

rzhade3 reviewed (changes requested):
Thanks for the PR and the added functionality! I have a couple comments below about the implementation, mainly to ensure that calling/ method signatures are consistent across the library.

Are you able to write and record tests for this additional functionality? If you don't have access to a test HackerOne instance, I can help record VCR tests, but would you please mind writing some tests, similar to [program_spec.rb](https://github.com/github/hackerone-client/blob/cd44f298bcccbafe8a4e77cbc0ff00d6ee8a9d84/spec/hackerone/client/program_spec.rb#L5) and [swag_spec.rb](https://github.com/github/hackerone-client/blob/cd44f298bcccbafe8a4e77cbc0ff00d6ee8a9d84/spec/hackerone/client/swag_spec.rb#L6)?

jeffgran-dox commented:
@rzhade3 thanks for the review! yes, I'll address your comments and work on adding some specs.

jeffgran-dox commented on lib/hackerone/client/organization.rb:
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
```
Get a program includes the `organization` object in the `relationships` part of the response: https://api.hackerone.com/customer-resources/#programs-get-program

Then the assets can be retrieved given the organization id via Get assets: https://api.hackerone.com/customer-resources/#assets-get-all-assets

Structured Scopes are also under the program: https://api.hackerone.com/customer-resources/#programs-get-structured-scopes

I initially tried to do what I was aiming for (updating scopes for my program) with the structured scopes API but the API returns an error saying "please use the new assets APIs instead" or something, so it seems like they are moving away from "scopes", at least for POST/PUT type operations. It's a little unclear to me still.

jeffgran-dox commented on lib/hackerone/client.rb:
```diff
@@ -4,10 +4,13 @@
 require "json"
 require "active_support"
 require "active_support/core_ext/numeric/time"
+require "ostruct"
```
I didn't need this when I was testing this locally but then when I tried to use it in another context it crashed with an "openstruct is not defined" error. I'm not sure what the cause was, maybe a different ruby version that doesn't require it by default any more or something. So I feel like it's worth leaving it in to prevent someone else from running into that, and like you say I think it should be harmless.

jeffgran-dox commented:
@rzhade3 Hi, I believe I have addressed your feedback and added some specs. However, I do not have an easy way to record new VCR fixtures so I just mocked the requests with webmock, pulling the example responses from their API docs. If you'd rather have these use VCR would you mind modifying the code and recording them for me? Thanks

rzhade3 reviewed (approved):
Thanks for the changes! I'm OK with using Webmock to stub out the request as opposed to re-recording the entire VCR request.

\cc @boveus for final review to see if there's anything more idiomatic we can do with Webmock

rzhade3 commented on spec/spec_helper.rb:
```diff
@@ -4,6 +4,7 @@
 require "hackerone/client"
 require "pry"
 require "vcr"
+require 'webmock/rspec'
```
```suggestion
require "webmock/rspec"
```

rzhade3 commented on spec/hackerone/client/asset_spec.rb:
```diff
@@ -0,0 +1,177 @@
+    JSON
+
+    stub_request(:get, "https://api.hackerone.com/v1/organizations/14/assets?page%5Bnumber%5D=1&page%5Bsize%5D=100").
+      to_return(body:<<~JSON2)
```
```suggestion
      to_return(body: <<~JSON2)
```

rzhade3 commented on lib/hackerone/client/asset.rb:
```diff
@@ -0,0 +1,60 @@
+      end
+
+      def programs
+        relationships.programs[:data].map{ |p| Program.new(p) }
```
```suggestion
        relationships.programs[:data].map { |p| Program.new(p) }
```

rzhade3 reviewed (commented):
Fixing Rubocop errors

rzhade3 commented:
@jeffgran-dox Thanks again for the PR, I've released these changes as version `0.23.0` on Rubygems.

jeffgran-dox commented:
@rzhade3 excellent, thank you!
